File: C++/src/VdSQLite.cpp
/*
 * VdSQLite.cpp
 *
 *  Created on: 29.11.2017
 *      Author: timm
 */

#include "VdSQLite.h"
#include <stdio.h>
#include <iostream>
#include <sqlite3.h>
using namespace std;

VdSQLite::VdSQLite(string filename) {
	// class for writing data to sqlite database
	this->open(filename);
	sqlite3 *db;
}

void VdSQLite::open(string filename) {
	/*
	 opens a new db file
	 :param filename: name of db file
	 :type filename: str
	 */
	char *zErrMsg = 0;
	filename = this->makeFilename(string("db"), filename);
	sqlite3_open(filename.c_str(), &this->db);

	string sql =
			"CREATE TABLE IF NOT EXISTS raw_data (id INTEGER PRIMARY KEY AUTOINCREMENT, time FLOAT, azimuth FLOAT, vertical FLOAT, distance FLOAT, reflection INTEGER)";
	sqlite3_exec(this->db, sql.c_str(), 0, 0, &zErrMsg);
	sqlite3_prepare_v2(db,
			"INSERT INTO raw_data (time, azimuth, vertical, distance, reflection) VALUES (?, ?, ?, ?, ?)",
			-1, &stmt, 0);
}

void VdSQLite::write() {
	// writes data to database
	char *zErrMsg = 0;
	sqlite3_exec(this->db, "BEGIN TRANSACTION;", 0, 0, &zErrMsg);
	for (VdPoint p : this->writingQueue) {
		sqlite3_bind_double(stmt, 1, p.getTime());
		sqlite3_bind_double(stmt, 2, p.getAzimuth());
		sqlite3_bind_double(stmt, 3, p.getVertical());
		sqlite3_bind_double(stmt, 4, p.getDistance());
		sqlite3_bind_int(stmt, 5, p.getReflection());
		sqlite3_step(stmt);
		sqlite3_reset(stmt);
	}
	sqlite3_exec(this->db, "END TRANSACTION;", 0, 0, &zErrMsg);
	this->clearWritingQueue();
}
void VdSQLite::close() {
	// close database
	sqlite3_close(db);
}

File: Python/vdSQLite.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
@author: Florian Timm
@version: 2017.11.27
"""

import sqlite3
from vdFile import VdFile
from vdPoint import VdPoint


class VdSQLite(VdFile):
    """ class for writing data to sqlite database """

    def _open(self, filename):
        """
        opens a new db file
        :param filename: name of db file
        :type filename: str
        """
        filename = self._make_filename("db", filename)
        print(filename)
        self.__db = sqlite3.connect(filename)
        self.__cursor = self.__db.cursor()
        self.__cursor.execute("CREATE TABLE IF NOT EXISTS raw_data ("
                              "id INTEGER PRIMARY KEY AUTOINCREMENT,"
                              "time FLOAT,"
                              "azimuth FLOAT,"
                              "vertical FLOAT,"
                              "distance FLOAT,"
                              "reflection INTEGER)")
        self.__db.commit()

    def write(self):
        """ writes data to database """
        insert = []
        for p in self.writing_queue:
            insert.append((p.time, p.azimuth, p.vertical,
                           p.distance, p.reflection))

        self.__cursor.executemany("INSERT INTO raw_data ("
                                  "time, azimuth, vertical, "
                                  "distance, reflection) "
                                  "VALUES (?, ?, ?, ?, ?)", insert)
        self.__db.commit()
        self.clear_writing_queue()

    def close(self):
        """ close database """
        self.__db.close()
